### PR TITLE
Add Latest Protocol Snapshot to Access API

### PR DESCRIPTION
**Pull request by danuio**
Resolves https://github.com/dapperlabs/flow-go/issues/4891.

### CODE REVIEW

**Review by jordanschalm — COMMENTED** [review]
Looks good, but can we add a test for the new method? Looks like we have tests set up for the `*Backend`, so maybe move more of the logic there (ie change the `*Backend` method to `GetLatestProtocolStateSnapshot(_ context.Context) ([]byte, error)` and handle the serialization there.

File: access/handler.go [hunk 1]
```diff
@@ -405,6 +408,28 @@ func (h *Handler) GetEventsForBlockIDs(
...
+
+	serializable, err := inmem.FromSnapshot(snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := json.Marshal(serializable)
+	if err != nil {
+		return nil, err
+	}
```
Comment (vishalchangrani) [inline access/handler.go:426]: I would suggest we put all of this in a new method in https://github.com/onflow/flow-go/blob/master/engine/common/rpc/convert/convert.go and then call that method. That would be inline with all other  calls.

File: engine/access/rpc/backend/backend.go [hunk 2]
```diff
@@ -185,6 +185,11 @@ func (b *Backend) GetNetworkParameters(_ context.Context) access.NetworkParamete
 	}
 }
 
+func (b *Backend) GetLatestProtocolStateSnapshot(_ context.Context) (protocol.Snapshot, error) {
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend.go:188]: may be add a simple happy path only unit test?

File: access/handler.go (hunk 1 above)
Comment (vishalchangrani) [reply access/handler.go:426]: `Convert.SnapshotToBytes(snapshot protocol.Snapshot)  ([]byte, error)`

File: engine/common/rpc/convert/convert.go [hunk 3]
```diff
@@ -359,3 +362,18 @@ func MessagesToIdentifiers(l [][]byte) []flow.Identifier {
 	}
 	return results
 }
+
+// SnapshotToBytes converts a `protocol.Snapshot` to bytes
```
Comment (jordanschalm) [inline engine/common/rpc/convert/convert.go:366]:
```suggestion
// SnapshotToBytes converts a `protocol.Snapshot` to bytes, encoded as JSON
```

File: engine/access/rpc/backend/backend_test.go [hunk 4]
```diff
@@ -122,6 +122,63 @@ func (suite *Suite) TestGetLatestFinalizedBlockHeader() {
...
+	phase := flow.EpochPhaseSetup
+	suite.snapshot.On("Phase").Return(phase, nil).Once()
+
+	epoch := protocol.EpochQuery{}
+	epoch.On("Current").Return(unittest.EpochSetupFixture(), nil)
+	epoch.On("Next").Return(unittest.EpochCommitFixture(), nil)
+	epoch.On("Preview").Return(unittest.EpochSetupFixture(), nil)
+
+	suite.snapshot.On("Epochs").Return(epoch, nil).Once()
+
```
Comment (jordanschalm) [inline engine/access/rpc/backend/backend_test.go:157]: Can simplify:
```go
snap := unittest.RootSnapshotFixture(unittest.CompleteIdentitySet())
suite.state.On("Sealed").Return(snap, nil).Once()
```

**Review by jordanschalm — COMMENTED** [review]
Nice 👏 

Once we update the mocking in `backend_test`, this LGTM

File: engine/common/rpc/convert/convert.go [hunk 5]
```diff
@@ -359,3 +362,18 @@ func MessagesToIdentifiers(l [][]byte) []flow.Identifier {
...
+func SnapshotToBytes(snapshot protocol.Snapshot) ([]byte, error) {
```
Comment (vishalchangrani) [inline engine/common/rpc/convert/convert.go:367]: 👍

**Review by vishalchangrani — APPROVED** [review]
lgtm - thanks for making those changes.

File: engine/access/rpc/backend/backend_test.go [hunk 6]
```diff
@@ -122,6 +122,35 @@ func (suite *Suite) TestGetLatestFinalizedBlockHeader() {
 	suite.Require().Equal(block.ParentID, header.ParentID)
 
 	suite.assertAllExpectations()
+
+}
+
+func (suite *Suite) TestGetLatestProtocolStateSnapshot() {
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_test.go:128]: 👍

**Review by vishalchangrani — APPROVED** [review]
lgtm!

**Review by jordanschalm — APPROVED** [review]

**Pull request merged**